File: packages/types/src/templates/corner-angle.ts
/**
 * Corner angle — підсилювальний кутник з auto-grid отворів.
 *
 * Геометрія = L-bracket (2 полиці під 90°, 1 гиб), але hole-pattern
 * — автоматичний grid rows×cols на кожній полиці, що генерується з
 * параметрів без ручних координат. Це типовий меблевий кутник
 * "купив і прикрутив".
 *
 *   ┌──────────────────────┐  ← horizontal flange (legB_mm × width_mm)
 *   │ ●     ●     ●        │     holes: rows × cols, distributed
 *   │ ●     ●     ●        │     evenly between hole_margin_mm edges
 *   └──────────────────────┘
 *   │
 *   │ ●     ●
 *   │ ●     ●  ← vertical flange (legA_mm × width_mm)
 *   │ ●     ●
 *   │
 *   └ bend (90°)
 *
 * Кількість отворів = 2 × hole_rows × hole_cols (по один patern на полицю).
 * MVP-обмеження ті ж самі: 90°, радіус з фіксованого набору.
 */
import { z } from "zod";

export const CornerAngleParametersSchema = z.object({
  /** Висота вертикальної полиці, мм. */
  legA_mm: z.number().min(20).max(500).describe("group:Полиця A|label:Висота полиці A (мм)"),
  /** Глибина горизонтальної полиці, мм. */
  legB_mm: z.number().min(20).max(500).describe("group:Полиця B|label:Глибина полиці B (мм)"),
  /** Внутрішній радіус гиба, мм. */
  bend_radius_mm: z
    .union([z.literal(1), z.literal(2.5), z.literal(4), z.literal(5)])
    .describe("group:Гиб|label:Внутрішній радіус (мм)"),
  /** MVP: лише 90°. */
  bend_angle_deg: z.literal(90).describe("group:Гиб|label:Кут гиба (°)"),
  /** Ширина (довжина гиба), мм. */
  width_mm: z.number().min(20).max(3000).describe("group:Загальне|label:Ширина (довжина гиба, мм)"),
  /** Діаметр отворів, мм (один для всього grid). */
  hole_diameter_mm: z
    .number()
    .min(3)
    .max(20)
    .describe("group:Сітка отворів|label:Діаметр отвору (мм)"),
  /** Рядів отворів вздовж ширини (perpendicular to leg length). */
  hole_rows: z.number().int().min(0).max(5).describe("group:Сітка отворів|label:Рядів отворів"),
  /** Колонок отворів вздовж довжини полиці (parallel to leg length). */
  hole_cols: z.number().int().min(0).max(5).describe("group:Сітка отворів|label:Колонок отворів"),
  /** Мінімальна відстань від країв полиці до центра крайнього отвору, мм. */
  hole_margin_mm: z
    .number()
    .min(5)
    .max(50)
    .describe("group:Сітка отворів|label:Відступ від країв (мм)"),
});

export type CornerAngleParameters = z.infer<typeof CornerAngleParametersSchema>;

export const CORNER_ANGLE_DEFAULT_PARAMETERS: CornerAngleParameters = {
  legA_mm: 50,
  legB_mm: 50,
  bend_radius_mm: 2.5,
  bend_angle_deg: 90,
  width_mm: 80,
  hole_diameter_mm: 5,
  hole_rows: 1,
  hole_cols: 2,
  hole_margin_mm: 12,
};

File: workers/cad/flatcraft_cad/templates/corner_angle.py
"""Corner angle — підсилювальний кутник з авто-grid отворів.

Геометрія = L-кронштейн (2 полиці, 1 90° гиб), але hole pattern
автоматично генерується з параметрів rows×cols×margin.

Призначення: меблевий/конструкційний кутник, де користувач не задає
координат — лише кількість отворів у grid'і.

Контракт параметрів збігається з `packages/types/templates/corner-angle.ts`.
"""

from typing import Literal

import cadquery as cq
from pydantic import BaseModel, ConfigDict, Field, field_validator

from flatcraft_cad.templates.base import BendDirection, Template

ALLOWED_INNER_RADIUS_MM: tuple[float, ...] = (1.0, 2.5, 4.0, 5.0)


class CornerAngleBuildParameters(BaseModel):
    """Усі параметри corner_angle для CadQuery-builder'у.

    Holes генеруються автоматично з rows/cols/margin → `compute_corner_angle_holes`.
    Координати JSON-aliases збігаються з TS-схемою (snake_case).
    """

    model_config = ConfigDict(frozen=True, populate_by_name=True)

    leg_a_mm: float = Field(
        ge=20, le=500, alias="legA_mm", description="Висота вертикальної полиці."
    )
    leg_b_mm: float = Field(
        ge=20, le=500, alias="legB_mm", description="Глибина горизонтальної полиці."
    )
    bend_radius_mm: float = Field(description="Внутрішній радіус гиба.")
    bend_angle_deg: Literal[90] = Field(default=90, description="MVP: тільки 90°.")
    bend_direction: BendDirection = Field(
        default="down", description="Напрям згину (Hotfix 2.10.e)."
    )
    width_mm: float = Field(ge=20, le=3000, description="Довжина лінії гиба.")
    thickness_mm: float = Field(gt=0, le=10, description="Товщина листа.")
    hole_diameter_mm: float = Field(ge=3, le=20, description="Діаметр отворів grid'у.")
    hole_rows: int = Field(ge=0, le=5, description="Рядів отворів вздовж ширини.")
    hole_cols: int = Field(ge=0, le=5, description="Колонок отворів вздовж полиці.")
    hole_margin_mm: float = Field(ge=5, le=50, description="Відступ від країв полиці.")

    @field_validator("bend_radius_mm")
    @classmethod
    def _radius_in_allowed_set(cls, value: float) -> float:
        if value not in ALLOWED_INNER_RADIUS_MM:
            raise ValueError(
                f"bend_radius_mm must be one of {list(ALLOWED_INNER_RADIUS_MM)}; got {value}"
            )
        return value


def build_corner_angle(params: CornerAngleBuildParameters) -> cq.Workplane:
    """3D corner_angle = L-profile, ідентичний L-кронштейну.

    Отвори не вирізаються у 3D-моделі (preview): для DXF/PDF вони
    обчислюються окремо через `compute_corner_angle_holes`. Це збігається
    зі стратегією L-bracket (web preview = solid extrude).
    """
    a = params.leg_a_mm
    b = params.leg_b_mm
    t = params.thickness_mm
    r = params.bend_radius_mm
    w = params.width_mm

    profile = (
        cq.Workplane("XZ")
        .moveTo(0, 0)
        .lineTo(b, 0)
        .lineTo(b, t)
        .lineTo(t + r, t)
        .radiusArc((t, t + r), -r)
        .lineTo(t, a)
        .lineTo(0, a)
        .close()
    )
    return profile.extrude(w)


class CornerAngleTemplate(Template[CornerAngleBuildParameters]):
    """Адаптер шаблонів — slug збігається з seed.ts."""

    name = "corner_angle"

    def build(self, params: CornerAngleBuildParameters) -> cq.Workplane:
        return build_corner_angle(params)
